agent/proxycfg/manager.go

```diff
@@ -0,0 +1,273 @@
+package proxycfg
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/cache"
+	"github.com/hashicorp/consul/agent/local"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// Manager is a component that integrates into the agent and manages Connect
+// proxy configuration state. This is used both for "managed" and "unmanaged"
+// proxies where "managed" refers to whether the _process_ itself is supervised.
+//
+// The Manager ensures that any Connect proxy registered on the agent has all
+// the state it needs cached locally via the agent cache. State includes
+// certificates, intentions, and service discovery results for any declared
+// upstreams.
+type Manager struct {
+	// cache is the agent's cache instance that can be used to retreive, store and
+	// monitor state for the proxies.
+	cache *cache.Cache
+	// state is the agent's local state to be watched for new proxy registrations.
+	state *local.State
+	// source describes the current agent's identity, it's used directly for
+	// prepared query discovery but also indirectly as a way to pass current
+	// Datacenter name into other request types that need it. This is sufficient
+	// for now and cleaner than passing the entire RuntimeConfig.
+	source *structs.QuerySource
+	// logger is the agent's logger to be used for logging logs.
+	logger *log.Logger
+
+	// stateCh is notified for any service changes in local state. We only use
+	// this to triger on _new_ service addition since it has no data and we don't
+	// want to maintain a full copy of the state in order to diff and figure out
+	// what changed. Luckily each service has it's own WatchCh so we can figure
+	// out changes and removals with those efficiently.
+	stateCh chan struct{}
+
+	mu       sync.Mutex
+	proxies  map[string]*State
+	watchers map[string]map[uint64]chan *ConfigSnapshot
+}
+
+// NewManager constructs a manager from the provided agent cache.
+func NewManager(logger *log.Logger, cache *cache.Cache, state *local.State,
+	source *structs.QuerySource) *Manager {
+	m := &Manager{
+		logger: logger,
+		cache:  cache,
+		state:  state,
+		source: source,
+		// Single item buffer is enough since there is no data transferred so this
+		// is "level triggering" and we can't miss actual data.
+		stateCh:  make(chan struct{}, 1),
+		proxies:  make(map[string]*State),
+		watchers: make(map[string]map[uint64]chan *ConfigSnapshot),
+	}
+	go m.run()
+	return m
+}
+
+// run is the internal state syncing loop that watches the local state for
+// changes.
+func (m *Manager) run() {
+
+	m.mu.Lock()
+	stateCh := m.stateCh
+	m.mu.Unlock()
+
+	if stateCh == nil {
```

Comment:
Why is this conditional here? Maybe just a comment explaining why/when this would ever be the case.

Reply:
Added comment.